reco_utils/recommender/tfidf/tfidf_utils.py

```diff
@@ -0,0 +1,301 @@
+        clean = clean.replace('\n', ' ')
+        clean = clean.replace('\t', ' ')
+        clean = clean.replace('\r', ' ')
+        clean = clean.replace('Â\xa0', '')     # non-breaking space
+
+        # Remove all punctuation and special characters
+        clean = re.sub('([^\s\w]|_)+','', clean)
+
+        # If you want to keep some punctuation, see below commented out example
+        # clean = re.sub('([^\s\w\-\_\(\)]|_)+','', clean)
+
+        # Skip further processing if the text will be used in BERT tokenization
+        if for_BERT is False:
+            # Lower case
+            clean = clean.lower()
+    except:
+        if verbose is True:
+            print('Cannot clean non-existent text')
+        clean = ''
+    
+    return clean
+
+def clean_dataframe_for_rec(df, cols_to_clean, for_BERT=False):
+    """ Clean the text within the columns of interest and return a dataframe with cleaned and combined text.
+    
+    Args:
+        df (pd.DataFrame): Dataframe to clean.
+        cols_to_clean (str): List of columns to clean by name (e.g., ['title','abstract','full_text']).
+        for_BERT (boolean): True or False for if this text is being cleaned for future BERT tokenization.
+    
+    Returns:
+        df (pd.DataFrame): Dataframe with cleaned text in new column 'cleaned_text'.
+    """
+    # Clean the text in the dataframe
+    for i in range(0, len(df)):
+        for col in cols_to_clean:
+            df[col][i] = clean_text(df[col][i], for_BERT)
+
+    # Collapse the table such that all descriptive text is just in a single column
+    first_col = True
+    for col in cols_to_clean:
+        if first_col is True:
+            # first one
+            df['cleaned_text'] = df[col]
+            first_col = False
+        else:
+            df['cleaned_text'] = df['cleaned_text'] + ' ' + df[col]
+    
+    # Make sure any punctuation or special characters in Name are human readible
+    for i in range(0, len(df)):
+        df['title'][i] = df['title'][i].encode('ascii','ignore')
+    
+    for i in range(0, len(df)):
+        df['title'][i] = df['title'][i].decode()
+    
+    return df
+
+def tokenize_with_BERT(vector, bert_method='bert-base-cased'):
+    """ Tokenize the input text with HuggingFace BERT tokenization.
+    
+    Args:
+        vector (pd.Series): Series of textual descriptions (typically values from pandas series df_clean['full_text']).
+    
+    Returns:
+        vector_tokenized (pd.Series): HuggingFace BERT tokenized input.
+    """
+    
+    # Load pre-trained model tokenizer (vocabulary)
+    tokenizer = BertTokenizer.from_pretrained(bert_method)
+    
+    # Loop through each item
+    vector_tokenized = vector.copy()
+    for i in range(0, len(vector)):
+        vector_tokenized[i] = ' '.join(tokenizer.tokenize(vector[i]))
+    
+    return vector_tokenized
+
+def recommend_with_tfidf(df_clean, text_col='cleaned_text', id_col='cord_uid', title_col='title', tokenization_method='scibert'):
```

Comment:
I think it would be good if you could refactor this part to make it more modular. Here one alternative:

One option is to create a class A=name_of_your_recommender and another class called Tokenizer. Class A would have an init with parameters related to the class and a tokenizer class. Then, there would be a fit method that would apply the tokenizer to the input text and that will fit tfidf. This code is a good example: https://github.com/microsoft/nlp-recipes/blob/master/utils_nlp/models/bert/sequence_encoding.py. Then, create another method to get the similar items. And the input potentially can be different metrics.

This is just an option, maybe there are some other ideas to make it more modular

Reply:
I have now created a class called `TfidfRecommender` to make it more modular.